components/feature/autofill/src/main/java/mozilla/components/feature/autofill/structure/ParsedStructureData.kt

```diff
@@ -0,0 +1,132 @@
+            .toTypedArray()
+    }
+
+    // Construct the saveInfo mask based upon the autofillIds that are available.
+    // This relies on the paired arrays of `saveDataTypeMasks` and the null padded `allPossibleIds`.
+    val saveInfoMask: Int by lazy {
+        allPossibleIds.mapIndexed { index, autofillId ->
+            autofillId?.let { saveDataTypeMasks[index] } ?: 0
+        }
+            .reduce { totalMask, saveDataTypeMask ->
+                totalMask or saveDataTypeMask
+            }
+    }
+
+    constructor(parcel: Parcel) : this(
+        parcel.readParcelable(AutofillId::class.java.classLoader),
+        parcel.readParcelable(AutofillId::class.java.classLoader),
+        parcel.readString(),
+        parcel.readString() ?: throw ParcelFormatException("packageName")
+    )
+
+    override fun writeToParcel(parcel: Parcel, flags: Int) {
+        parcel.writeParcelable(this.usernameId, flags)
+        parcel.writeParcelable(this.passwordId, flags)
+        parcel.writeString(this.webDomain)
+        parcel.writeString(this.packageName)
+    }
+
+    override fun describeContents(): Int {
+        return 0
+    }
+
+    companion object CREATOR : Parcelable.Creator<ParsedStructure> {
+        override fun createFromParcel(parcel: Parcel): ParsedStructure {
+            return ParsedStructure(parcel)
+        }
+
+        override fun newArray(size: Int): Array<ParsedStructure?> {
+            return arrayOfNulls(size)
+        }
+    }
+}
+
+internal open class ParsedStructureData<Id>(
+    val usernameId: Id? = null,
+    val passwordId: Id? = null,
+    val webDomain: String? = null,
+    val packageName: String
+) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is ParsedStructureData<*>) return false
+
+        if (usernameId != other.usernameId) return false
+        if (passwordId != other.passwordId) return false
+        if (webDomain != other.webDomain) return false
+        if (packageName != other.packageName) return false
+
+        return true
+    }
+
+    @Suppress("MagicNumber")
+    override fun hashCode(): Int {
+        var result = usernameId?.hashCode() ?: 0
+        result = 31 * result + (passwordId?.hashCode() ?: 0)
+        result = 31 * result + (webDomain?.hashCode() ?: 0)
+        result = 31 * result + packageName.hashCode()
+        return result
+    }
+}
+
+// This should be kept in the same order as `allPossibleIds` below.
+@RequiresApi(Build.VERSION_CODES.O)
+private val saveDataTypeMasks = arrayOf(
+    SaveInfo.SAVE_DATA_TYPE_USERNAME,
+    SaveInfo.SAVE_DATA_TYPE_PASSWORD
+)
+
+@RequiresApi(Build.VERSION_CODES.O)
+internal fun parseStructure(context: Context, structure: AssistStructure): ParsedStructure? {
+    val activityPackageName = structure.activityComponent.packageName
+    if (context.packageName == activityPackageName) {
+        // We do not autofill our own activities. Browser content will be auto-filled by Gecko.
```

Comment:
Does android autofill require display of some kind of ui? if not and it can work without a direct user interaction, i wonder if we can skip gecko's autofill and just rely on `feature-autofill` entirely...

Reply:
Right now the implementation always shows an account popup for selecting the account to autofill. I'm not sure if that is mandatory - would need to read the docs.
However for web content I think I'd still prefer Gecko to do it: The teams implementing that will do a better job identifying and filling web content. Piping this through GeckoView and then implementing our own heuristics will not beat that, I assume. I also think I remember that the GeckoView autofill implementation was very basic, so I guess we may not even have all the information to autofill more complex scenarios.